brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,171 @@
+  }
+
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIG_IGN;
+  sigaction(SIGPIPE, &action, NULL);
+
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+}
+#endif
+
+namespace brave {
+
+class TorLauncherDelegate : public base::LaunchOptions::PreExecDelegate {
+ public:
+    TorLauncherDelegate() {}
+    ~TorLauncherDelegate() override {}
+
+    void RunAsyncSafe() override {
+#if defined(OS_POSIX)
+      for (size_t i = 0; i < 2; ++i)
+        close(pipehack[i]);
+#endif
+      // TODO(darkdh): doesn't prevent ctrl+c propagated by terminal
+      struct sigaction action;
+      memset(&action, 0, sizeof(action));
+      action.sa_handler = SIG_IGN;
+      sigaction(SIGINT, &action, NULL);
+    }
+ private:
+    DISALLOW_COPY_AND_ASSIGN(TorLauncherDelegate);
+};
+
+TorLauncherImpl::TorLauncherImpl(
+    std::unique_ptr<service_manager::ServiceContextRef> service_ref)
+    : service_ref_(std::move(service_ref)) {
+  SetupPipeHack();
+}
+
+TorLauncherImpl::~TorLauncherImpl() {
+  if (tor_process_.IsValid()) {
+#if defined(OS_POSIX)
+    for (size_t i = 0; i < 2; ++i)
+      close(pipehack[i]);
+#endif
+    tor_process_.Terminate(0, false);
+#if defined(OS_MACOSX)
+    base::PostTaskWithTraits(
+        FROM_HERE, {base::MayBlock(), base::TaskPriority::BACKGROUND},
+        base::BindOnce(&base::EnsureProcessTerminated, Passed(&tor_process_)));
+#else
+    base::EnsureProcessTerminated(std::move(tor_process_));
+#endif
+  }
+}
+
+void TorLauncherImpl::Launch(const base::FilePath& tor_bin,
+                             const std::vector<std::string>& args,
+                             LaunchCallback callback) {
+  base::CommandLine cmdline(tor_bin);
+  for (size_t i = 0; i < args.size(); ++i) {
+    cmdline.AppendArg(args[i]);
+  }
+  base::LaunchOptions launchopts;
+  TorLauncherDelegate tor_launcher_delegate;
+  launchopts.pre_exec_delegate = &tor_launcher_delegate;
+#if defined(OS_LINUX)
+  launchopts.kill_on_parent_death = true;
+#endif
+  tor_process_ = base::LaunchProcess(cmdline, launchopts);
+
+  bool result;
+  // TODO(darkdh): return success when tor connected to tor network
+  if (tor_process_.IsValid())
+    result = true;
+  else
+    result = false;
+
+  if (callback)
+    std::move(callback).Run(result);
+
+  if (!child_monitor_thread_.get()) {
+    child_monitor_thread_.reset(new base::Thread("child_monitor_thread"));
+    if (!child_monitor_thread_->Start()) {
+      NOTREACHED();
+    }
+
+    child_monitor_thread_->task_runner()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&TorLauncherImpl::MonitorChild, base::Unretained(this)));
+  }
+}
+
+void TorLauncherImpl::SetCrashHandler(SetCrashHandlerCallback callback) {
+  crash_handler_callback_ = std::move(callback);
+}
+
+void TorLauncherImpl::MonitorChild() {
+#if defined(OS_POSIX)
+  pid_t pid;
+
+  while (1) {
+      if (read(pipehack[0], &pid, sizeof(pid_t)) > 0) {
```

Comment:
Can you add a compile-time assertion that sizeof(pid_t) < PIPE_BUF?  Under that assumption we are guaranteed by POSIX that the read/write will be atomic and we don't need to split it into potentially multiple reads/writes.

Alternatively, the SIGCHLD handler could just write a single byte to the pipe, and the thread could call waitpid.

Reply:
Actually on second thought I'm not sure this is guaranteed.  All that is guaranteed, I think, is that the write will fail or succeed completely without any concurrent write interposed in the middle.

So, simpler if the SIGCHLD handler just writes a single byte, and the thread reads as many bytes as are available without blocking as soon as is ready.  Something like this:

```
while (1) {
    // Wait for signal.
    fd_set fds;
    FD_ZERO(&fds);
    FD_SET(&fds, pipehack[0]);
    if (select(pipehack[0] + 1, &fds, NULL, NULL, NULL, NULL) == -1) {
        LOG(ERROR) << ...;
    }
    if (!FD_ISSET(&fds, pipehack[0]))
        continue;  // spurious wakeup

    // Drain signal pipe.
    char buf[PIPE_BUF];
    while (read(pipehack[0], buf, sizeof buf) > 0)
        continue;

    // Reap children.
    int status;
    pid_t pid;
    while ((pid = waitpid(-1, &status, WNOHANG)) != 0) {
        if (pid == -1) {
            LOG(ERROR) << ...;
            break;
        }
        ...examine status, decide what to do...
    }
}
```

Reply:
I actually set reading end as BLOCK and only set writing end as NON_BLOCK

Reply:
Hmm...  It may be that if the read end of the pipe is set to blocking, then the select is unnecessary.  I still recommend calling waitpid in the thread rather than signal handler!

Reply:
fixed in fea7dca